### PR TITLE
Add/34 create new category field modal

### LINKED ISSUE #35173: Category management improvements
**Description:** the new category management flow has some flaws that should be fixed before we test the new experience with users

**Acceptance criteria:**

- [x] The user should see an option to create a new category when he starts typing
- [x] The category list should appear on the first click

### PR DESCRIPTION
**Pull request by louwie17**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Partialy closes https://github.com/woocommerce/woocommerce/issues/35173

Adds a create category option to the categories dropdown when the search value does not exist.

https://user-images.githubusercontent.com/2240960/197212966-a646a474-1c14-4728-932a-5b07f59b6a9d.mp4

### How to test the changes in this Pull Request:

1. Load this branch, build it, and make sure you have the `new-product-management-experience` feature flag enabled (you can do so using the latest version of the Beta tester within the mono repo).
2. Go to **Products > Add New (MVP)**
3. Start typing in the search value of the Categories dropdown
4. If the search value you typed does not match an exact category a `+ Create 'category name'` option should show
5. Click the create option and a Modal should show with the category name pre-filled with the search value.
6. You have the option to select a parent (this is a single select dropdown in a similar format as the category dropdown)
7. Once you click **Save** it should create the category and add it as a selected value to the Categories list.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce-admin/client/products/fields/category-field/category-field-item.tsx [hunk 1]
```diff
@@ -89,7 +87,7 @@ export const CategoryFieldItem: React.FC< CategoryFieldItemProps > = ( {
 				<CheckboxControl
 					label={ item.data.name }
 					checked={ selectedIds.includes( item.data.id ) }
-					onChange={ () => item.data /*&& onSelect( item.data )*/ }
```
Comment (louwie17) [inline plugins/woocommerce-admin/client/products/fields/category-field/category-field-item.tsx:92]: Don't need the `onSelect` anymore as we are making using of the `getItemProps` props that handles this.

File: plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+			<div className="woocommerce-create-new-category-modal__wrapper">
+				<TextControl
+					label={ __( 'Name', 'woocommerce' ) }
+					name="Tops"
+					value={ categoryName }
+					onChange={ setCategoryName }
+				/>
+				<SelectControl< Pick< ProductCategory, 'id' | 'name' > >
+					items={ categoriesSelectList }
+					label={ __( 'Parent category (optional)', 'woocommerce' ) }
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx]: Made a similar comment [here](https://github.com/woocommerce/woocommerce/pull/35266#discussion_r1003839238) but we might consider moving out `(optional)` to make it easier on translators.

File: plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+				name: categoryName,
+				parent: categoryParent ? categoryParent.id : undefined,
+			} );
+			invalidateResolutionForStoreSelector( 'getProductCategories' );
+			setIsCreating( false );
+			onCreated( newCategory );
+		} catch ( e ) {
+			createNotice(
+				'error',
+				__( 'Failed to create category.', 'woocommerce' )
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx:73]: I know error handling wasn't the specific focus of this PR, but long-term we need to come up with better error handling around more common errors across all product components.

Duplicate slugs are a common one that I've been hitting on accident when testing and we need more specific error messages to help users navigate these issues.

File: plugins/woocommerce-admin/client/products/fields/category-field/category-field-add-new-item.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+> = ( { item, highlightedIndex, getItemProps, items } ) => {
+	const index = items.findIndex( ( i ) => i.id === item.id );
+	return (
+		<li
+			{ ...getItemProps( {
+				item,
+				index,
+			} ) }
+			className={ classNames(
+				'woocommerce-category-field-dropdown__item woocommerce-category-field-dropdown__new-item',
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/category-field-add-new-item.tsx]: Nitpicky, but this has 2 class names used as identifiers.  Should one of these be a modifier?  E.g., `woocommerce-category-field-dropdown__item is-new`

File: plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+/**
+ * Internal dependencies
+ */
+import './create-category-modal.scss';
+import { useCategorySearch } from './use-category-search';
+import { CategoryFieldItem } from './category-field-item';
+
+type CreateCategoryModalProps = {
+	initialCategoryName?: string;
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx:28]: Nice prop!

File: plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+	onCancel: () => void;
+	onCreated: ( newCategory: ProductCategory ) => void;
+};
+
+export const CreateCategoryModal: React.FC< CreateCategoryModalProps > = ( {
+	initialCategoryName,
+	onCancel,
+	onCreated,
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx]: Should we rename this to `onCreate` to keep tense consistent for callbacks?

File: plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+			onCancel();
+		}
+	};
+
+	const onInputChange = ( searchString?: string ) => {
+		searchCategories( searchString || '' );
+	};
+
+	const searchDelayed = useMemo(
+		() => debounce( onInputChange, 150 ),
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx]: No big deal, but noticing we have 150, 200, and 250 as debounce values.  No strong opinions on this, but wonder if we should make them somewhat consistent 🤔

File: plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+													/>
+												);
+											} ) }
+								</>
+							</Menu>
+						);
+					} }
+				</SelectControl>
+				<div className="woocommerce-create-new-category-modal__buttons">
+					<Button isSecondary onClick={ () => onCancel() }>
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.tsx]: Should this button also be disabled when `isCreating`?

**Review by joshuatf — COMMENTED** [review]
Super awesome functionality and very exciting to see this come to life!  Excellent work on this.

Tested pretty well for me, just noted some minor nitpick comments and 2 small issues with functionality:

1. When no parent categories are found based on query, I still see the border for the options menu:
Screen Shot 2022-11-01 at 4 44 17 PM

2. When I don't select a parent category, I consistently get the "Failed to create category" error.

**Comment by louwie17** [comment]
Thanks @joshuatf for the thorough review :) appreciate it, I have updated everything, mostly all as part of this commit: https://github.com/woocommerce/woocommerce/pull/35132/commits/7b4936a61576cbb30c135272b88bfdfd4ea18485

Would appreciate a re-review.

File: plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.scss [hunk 9]
```diff
@@ -0,0 +1,33 @@
+.woocommerce-create-new-category-modal {
+	min-width: 650px;
+
+	&__buttons {
+		margin-top: $gap-larger;
+		display: flex;
+		flex-direction: row;
+		gap: 8px;
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.scss]:
```suggestion
		gap: $gap-smaller;
```

File: plugins/woocommerce-admin/client/products/fields/category-field/category-field-add-new-item.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+type CategoryFieldAddNewItemProps = {
+	item: Pick< ProductCategory, 'id' | 'name' >;
+	highlightedIndex: number;
+	items: Pick< ProductCategory, 'id' | 'name' >[];
+} & Pick<
+	MenuItemProps< Pick< ProductCategory, 'id' | 'name' > >,
+	'getItemProps'
+>;
+
+export const CategoryFieldAddNewItem: React.FC<
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/fields/category-field/category-field-add-new-item.tsx:20]: I really like how this is separated into its own component 💯

**Review by joshuatf — DISMISSED** [review]
Testing wonderfully!  Thanks for all the changes and sorry for the delay in review on this one.

Left one minor comment on using a gap variable, but pre-approving ✅

File: plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.scss (hunk 9 above)
Comment (louwie17) [reply plugins/woocommerce-admin/client/products/fields/category-field/create-category-modal.scss]: Done as part of https://github.com/woocommerce/woocommerce/pull/35132/commits/df8f59cae1c3f98db90dfc6d4055e1aa1568d204

**Review by joshuatf — APPROVED** [review]
Thanks for the variable update! 🚢

**Pull request merged**